src/reference-manual/user-functions.Rmd

```diff
@@ -36,10 +36,42 @@ real foo(real mu, real sigma);
 
 declares a function named `foo` with two argument variables of
 types `real` and `real`.  The arguments are named `mu`
-and `sigma`, but that is not part of the declaration.  Two
-user-defined functions may *not* have the same name even if they have
-different sequences of argument types.
+and `sigma`, but that is not part of the declaration.
 
+### Function overloading
+Two user-defined functions may have the same name if they have
+different sequences of [argument types](#argument-types-and-qualifiers).
+
+This is known as function overloading.
+
+For example, the following two functions are both defined with the name `add_up`
+
+```stan
+real add_up(real a, real b){
+  return a + b;
+}
+
+real add_up(real a, real b, real c){
+  return a + b + c;
+}
+```
+
+The return types of overloaded functions do not need to be the same as a matter
+of course, but it may be confusing if they differ.
+
+Two function may **not** use the same name if their signature _only_ differs by the
+return type.
+
+For example,
+```stan
+real baz(int x);
+int baz(int x);
+```
+is not permitted.
+
+Function names used in the Stan standard library may be overloaded by
+user-defined functions. Exceptions to this are the `reduce_sum` family of
+functions and ODE integrators, which cannot be overloaded.
```

Comment:
This is not strictly true - users can override `map_rect`, for example. 

The true restriction is actually that UDFs cannot override variadic functions, as these are currently typechecked entirely separately based on their name. I can say this if we want, but I was trying to avoid the word "variadic"